tests/lib/tools/tests.device-cgroup

```diff
@@ -0,0 +1,210 @@
+    in the correct way"""
+    group_name = "snap.{}".format(snap_app)
+    with open(
+        os.path.join("/sys/fs/cgroup/devices/", group_name, "devices.list")
+    ) as inf:
+        sys.stdout.write(inf.read())
+
+
+def device_change_v1(snap_app, allow, kind, major_minor):
+    group_name = "snap.{}".format(snap_app)
+    devices_control = "devices.allow"
+    if not allow:
+        devices_control = "devices.deny"
+    with open(
+        os.path.join("/sys/fs/cgroup/devices/", group_name, devices_control), "w+"
+    ) as outf:
+        outf.write("{} {} rwm".format(kind, major_minor))
+
+
+class DeviceCgroupV2Key(ctypes.Structure):
+    """A class representing v2 device cgroup key entry, which is 9 bytes long"""
+
+    # UINT32_MAX
+    DEVICE_MINOR_ANY = 4294967295
+
+    _pack_ = 1
+    _fields_ = [
+        # type is already a correct v1-like char 'b', 'c'
+        ("_typ", ctypes.c_char),
+        ("major", ctypes.c_uint32),
+        ("minor", ctypes.c_uint32),
+    ]
+
+    @property
+    def typ(self):
+        return self._typ.decode()
+
+
+def dump_v2(snap_app):
+    """Dump device entries for a v2 device cgroup. Entries are stored in a BPF map
+    created by snap-confine. We dump its contents using bpftool, and then
+    convert each entry to a format that looks like v1.
+    """
+    snap, app = snap_app.split(".")
+    map_pin_name = "snap_{}_{}".format(snap, app)
+    output = subprocess.check_output(
+        [
+            "bpftool",
+            "map",
+            "dump",
+            "pinned",
+            "/sys/fs/bpf/snap/{}".format(map_pin_name),
+            "-j",
+        ],
+    )
+    logging.debug("got output %s", output)
+    entries = json.loads(output)
+    for entry in entries:
+        raw_key = entry.get("key")
+        if not raw_key:
+            raise RuntimeError("unexpected object content in {}".format(output))
+        if len(raw_key) != ctypes.sizeof(DeviceCgroupV2Key):
+            raise RuntimeError(
+                "unexpected size of raw key {} (expected {})".format(
+                    len(raw_key), ctypes.sizeof(DeviceCgroupV2Key)
+                )
+            )
+        raw = bytearray([int(v, 16) for v in raw_key])
+        key = DeviceCgroupV2Key.from_buffer(raw)
+        # minor can be either the actual value, or a special value '*' which
+        # represent any minor number
+        minor = key.minor
+        if minor == DeviceCgroupV2Key.DEVICE_MINOR_ANY:
+            minor = "*"
+        print(
+            "{typ} {major}:{minor} rwm".format(
+                typ=key.typ,
+                major=key.major,
+                minor=minor,
+            )
+        )
+
+
+def device_change_v2(snap_app, allow, kind, major_minor):
+    """Dump device entries for a v2 device cgroup. Entries are stored in a BPF map
+    created by snap-confine. We dump its contents using bpftool, and then
+    convert each entry to a format that looks like v1.
```

Comment:
I guess this docstring should be updated, since it's the same as in the previous function.